define defaults for context propagation

njr-11 opened the pull request:
fixes #27

Add a constant for context propagation defaults per discussion under #27

Signed-off-by: Nathan Rauh <nathan.rauh@us.ibm.com>

tomjenkinson commented on api/src/main/java/org/eclipse/microprofile/concurrent/ManagedExecutorConfig.java:
```diff
@@ -65,7 +69,7 @@
      * action or task.</p>
      */
-    String[] propagated() default { ThreadContext.APPLICATION, ThreadContext.CDI, ThreadContext.SECURITY };
+    String[] propagated() default { ThreadContext.DEFAULTS };
```
In my opinion this is a bit confusing (it to remove the explicit set of defaults in the Java code and rely on readers to the Javadoc). It seems to be similar opinion of author of comment https://github.com/eclipse/microprofile-concurrency/issues/27#issuecomment-430120816 @manovotn

manovotn replied:
Yes, but the PR now basically states that all of the context excepting transaction are propagated.
That's IMO good - what I didn't like was the previous notion of leaving default undefined and based on implementation.

njr-11 replied:
@tomjenkinson there are a couple of problems that are causing us to move away from the approach of enumerating an exact list of individual contexts as the default.

1. it makes the default ManagedExecutor/ThreadContext unusable if any one of the types listed isn't supported by the product or has been intentionally disabled by the user.  For example, if a product doesn't include CDI or the user turns off CDI, then the basic @Inject ManagedExecutor will be unusable per the direction we are headed under #24.
2. It makes for a suboptimal user experience when a container or third party provides any additional context types.  The user would miss out on propagation of these types by default.

Possibly, DEFAULTS isn't the best name for this, and I'm open to improvements that anyone has to offer.  What it really means to say is either
"all available contexts except for transaction context, which is instead suspended/cleared"
or
"all available contexts which say they should be default where transaction context must claim it is not default and is instead suspended/cleared".
There is an ongoing discussion between the above options on the meaning for DEFAULTS.
We also have the ALL constant for all available contexts _including_ transactions, which could be trivially replaced by (DEFAULTS, TRANSACTION) if the former meaning is chosen.

Based on your comment, I can guess that you would favor the former meaning here.  In that case, would it be helpful to rename DEFAULTS to ALL_EXCEPT_TRANSACTION ?

tomjenkinson replied:
Given what you said it sounds a bit more like a blacklist rather than a whitelist (as those in the whitelist my not be available for the reasons you mentioned).
Maybe:
`String[] defaultNotPropagated() { ThreadContext.TRANSACTION }`

njr-11 replied:
@tomjenkinson The name `defaultNotPropagated` doesn't make sense to me (a default isn't a configurable attribute that a user sets to various values), but I understand the general idea here, and it is a good idea.

There are three categories of context propagation (propagated, unchanged, cleared), but only the first two of these are currently configurable, with the third being implied by absence of a context type in the other sets.  However, we could do this differently and allow the explicit configuration of all three sets.  It would be redundant, but more straightforward, and I think probably worth it.

In this case, we would define the ALL constant to mean all available context types that aren't otherwise individually listed elsewhere.  Would ALL_OTHER be a more accurate name? Does anyone have preferences on that?

So, for ThreadContext, we would have:

```
String[] propagated() default { ThreadContext.ALL };
String[] cleared() default { ThreadContext.TRANSACTION };
String[] unchanged() default {};
```

And for ManagedExecutor, we would have:

```
String[] propagated() default { ThreadContext.ALL };
String[] cleared() default { ThreadContext.TRANSACTION };
```
* Note that `unchanged` is intentionally forbidden on ManagedExecutor because it would introduce the same non-deterministic, unsafe, unpredictable behavior that we wrote the spec to solve.

There would no longer be a need for a DEFAULTS constant because it would already be the default.

tomjenkinson replied:
It would need to be clear in the text of the specification that cleared overrides propagated but yes. It some ways would seem to be a bit like how the excludes annotation is used in the Ant fileset example: https://ant.apache.org/manual/Types/fileset.html

`<fileset dir="${server.src}" casesensitive="yes">
  <include name="**/*.java"/>
  <exclude name="**/*Test*"/>
</fileset>`

njr-11 replied:
It isn't that cleared overrides propagated.  Consider the opposite example.  I think this just made the case for not naming it ALL and calling it ALL_OTHER instead,

```
@ThreadContextConfig(propagated = ThreadContext.APPLICATION,
                     cleared = ThreadContext.ALL_OTHER)

@ThreadContextConfig(cleared = TRANSACTION,
                     propagated = ThreadContext.ALL_OTHER)
```

tomjenkinson replied:
I think it may be use the way I expressed override. If the first example means propagate everything but ThreadContext.APPLICATION and the second one is to propagate everything but TRANSACTION it looks good to me.

tomjenkinson replied:
I edited my previous comment above^

njr-11 commented:
@tomjenkinson Thanks for your comments. I think they have resulted in improving the design and making the configuration more intuitive.  I have added a commit that covers we discussed in your review comment thread.  Please review again. Thanks!
@manovotn and @FroMage please review as well.

njr-11 commented:
On the MP Concurrency call we thought of ALL_REMAINING as a better name than ALL_OTHER.  We also decided that we will leave this pull open for several days in case anyone thinks of better names to consider.  After that, I'll add another commit that updates to the chosen name.

To recap the discussion, there are 3 ways of dealing with thread context (all configurable).
Thread context can be **propagated** from the requesting thread to the thread that runs the operation.
Thread context can be **cleared** from thread that executes the action.
Thread context can remain **unchanged** on the thread that executes the action. (This one is only applicable when using ThreadContext to run actions inline, not when using a ManagedExecutor)

Let's say that someone wants a ThreadContext instance that clears all context when running actions, except for Transaction context, which it wants unchanged on the execution thread, and Security context, which it wants propagated from the requesting thread.
Then they would configure,
```
@Inject @ThreadContextConfig(propagated=ThreadContext.SECURITY,
                             unchanged=ThreadContext.TRANSACTION,
                             cleared=ThreadContext.ALL_REMAINING)
ThreadContext threadCtx;
```

So this last question is just over whether name ALL_REMAINING is good and sufficiently intuitive for describing this, or whether someone thinks of a better name for it.

njr-11 commented:
I haven't seen any suggestions come through for better names, so I have added a commit updating the constant name from ALL_OTHER to ALL_REMAINING as discussed on the MP Concurrency call last Friday.   I have also rebased the commits onto the latest from master, and included a couple of other minor JavaDoc fixes that I missed in the original pull.

@manovotn or @tomjenkinson could you approve this latest commit?
